hasher/tests/merkle.rs

```diff
@@ -0,0 +1,51 @@
+use ark_ff::Zero;
+use mina_hasher::*;
+use std::sync::RwLock;
+
+lazy_static::lazy_static! {
+    static ref DOMAIN_STRING_CALL_COUNTER: RwLock<usize> = RwLock::new(0);
+}
+
+#[derive(Debug, Clone)]
+struct TestMerkleNode {
+    height: u32,
+    left: Fp,
+    right: Fp,
+}
+
+impl Hashable for TestMerkleNode {
+    type D = ();
+
+    fn to_roinput(&self) -> ROInput {
+        let mut roi = ROInput::new();
+        roi.append_field(self.left);
+        roi.append_field(self.right);
+        roi
+    }
+
+    fn domain_string(this: Option<&Self>, _: Self::D) -> Option<String> {
+        match this {
+            None => format!("Unused").into(),
+            Some(x) => {
+                let mut counter = DOMAIN_STRING_CALL_COUNTER.write().unwrap();
+                *counter += 1;
+                format!("TestMerkleNode{:03}", x.height).into()
+            }
+        }
+    }
+}
+
+#[test]
+fn ensure_domain_string_is_invoked() {
+    let node = TestMerkleNode {
+        height: 3,
+        left: Fp::zero(),
+        right: Fp::zero(),
+    };
+    for i in 0..10 {
+        let mut hasher = create_legacy(());
```

Comment:
Yeah, it's not intended to be used like this.  As the examples show, you're force to init the hasher with something for domain separation.  For example, start with 0.

Then if you're forced to do some "weird" hashing, such as using the height as domain separation before each hash, then you need to use the right interface, which is not the `update` interface.

Instead you should use one of the one-shot interfaces, such as

```rust
let data = Example { height: 82, left: Fp::zero(), right: Fp::zero() };
let out = hasher.hash(&data);

// Or if you define domain_string without using `this`... like this
let out = hasher.init(data.height).update(&data).digest();
let out = hasher.init_and_hash(data.height, &data);
```

The update interface is for incremental stuff, so the pattern would have to be

```
init // sets domain separation
update
update
...
update
digest

init // sets another domain separation
update
update
...
update
digest
```

Update must definitely not be setting domain separation each time!

Reply:
`init` works except it resets the initial states, and in my case that prevents proper pooling, should there be an `init_lite` fn?

Reply:
It's not necessary to get what you want.  You can always call `init` to set the domain separation to whatever you want, whenever you want (e.g. after popping from the pool).

For example,

```
1. Setup merkle_hasher_pool with hashers domain separation initialised to whatever height (e.g. choose `0`)
2. hasher <- merkle_hasher_pool // grab hasher from pool
3. hasher.init(height) // set domain separation for height that you like
4. out <- hasher.hash(merkle_node)
5. put hasher back in the pool
```

Reply:
That works thanks!